arangod/Graph/Enumerators/OneSidedEnumerator.cpp

```diff
@@ -0,0 +1,266 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2021 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+/// @author Heiko Kernbach
+////////////////////////////////////////////////////////////////////////////////
+
+#include "OneSidedEnumerator.h"
+
+#include "Basics/debugging.h"
+#include "Basics/system-compiler.h"
+
+#include "Futures/Future.h"
+#include "Graph/Options/OneSidedEnumeratorOptions.h"
+#include "Graph/PathManagement/PathStore.h"
+#include "Graph/PathManagement/PathStoreTracer.h"
+#include "Graph/PathManagement/PathValidator.h"
+#include "Graph/Providers/ClusterProvider.h"
+#include "Graph/Providers/ProviderTracer.h"
+#include "Graph/Providers/SingleServerProvider.h"
+#include "Graph/Queues/FifoQueue.h"
+#include "Graph/Queues/QueueTracer.h"
+
+#include <Graph/Types/ValidationResult.h>
+#include <Logger/LogMacros.h>
+#include <velocypack/Builder.h>
+#include <velocypack/velocypack-aliases.h>
+
+using namespace arangodb;
+using namespace arangodb::graph;
+
+template <class QueueType, class PathStoreType, class ProviderType, class PathValidator>
+OneSidedEnumerator<QueueType, PathStoreType, ProviderType, PathValidator>::OneSidedEnumerator(
+    ProviderType&& forwardProvider, OneSidedEnumeratorOptions&& options,
+    arangodb::ResourceMonitor& resourceMonitor)
+    : _options(std::move(options)),
+      _queue(resourceMonitor),
+      _provider(std::move(forwardProvider)),
+      _validator(_interior),
+      _interior(resourceMonitor),
+      _resultPath{_provider, _provider} {}
```

Comment:
We only need 1 Provider, I am confused.

Reply:
Implemented new ResultType (which only needs one provider)